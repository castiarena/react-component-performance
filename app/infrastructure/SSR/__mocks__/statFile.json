{
  "errors": [],
  "warnings": [],
  "version": "4.44.2",
  "hash": "d9331d98e1777d265b96",
  "publicPath": "/public",
  "outputPath": "/Users/agustin.castiarena/apps/react-component-performance/dist/public",
  "assetsByChunkName": {
    "mock": [
      "mock.js",
      "mock.js.map"
    ],
    "vendors~mock": [
      "vendors~mock.bundle.js",
      "vendors~mock.bundle.js.map"
    ]
  },
  "assets": [
    {
      "name": "mock.js",
      "size": 3419,
      "chunks": [
        0
      ],
      "chunkNames": [
        "mock"
      ],
      "info": {},
      "emitted": false
    },
    {
      "name": "mock.js.map",
      "size": 10312,
      "chunks": [
        0
      ],
      "chunkNames": [
        "mock"
      ],
      "info": {
        "development": true
      },
      "emitted": false
    },
    {
      "name": "vendors~mock.bundle.js",
      "size": 170140,
      "chunks": [
        1
      ],
      "chunkNames": [
        "vendors~mock"
      ],
      "info": {},
      "emitted": false
    },
    {
      "name": "vendors~mock.bundle.js.map",
      "size": 426928,
      "chunks": [
        1
      ],
      "chunkNames": [
        "vendors~mock"
      ],
      "info": {
        "development": true
      },
      "emitted": false
    }
  ],
  "filteredAssets": 0,
  "entrypoints": {
    "mock": {
      "chunks": [
        1,
        0
      ],
      "assets": [
        "vendors~mock.bundle.js",
        "vendors~mock.bundle.js.map",
        "mock.js",
        "mock.js.map"
      ],
      "children": {},
      "childAssets": {}
    }
  },
  "namedChunkGroups": {
    "mock": {
      "chunks": [
        1,
        0
      ],
      "assets": [
        "vendors~mock.bundle.js",
        "vendors~mock.bundle.js.map",
        "mock.js",
        "mock.js.map"
      ],
      "children": {},
      "childAssets": {}
    }
  },
  "logging": {
    "webpack.buildChunkGraph.visitModules": {
      "entries": [],
      "filteredEntries": 2,
      "debug": false
    }
  },
  "children": []
}