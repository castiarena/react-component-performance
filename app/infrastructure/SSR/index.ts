import { createElement, ReactElement } from 'react';
import { renderToString } from 'react-dom/server';
import { ServerStyleSheet } from 'styled-components';
import Layout from '../../components/Layout';

class SSR {
  private sheet = new ServerStyleSheet();
  private readonly Component: any;

  constructor(Component: any) {
    this.Component = Component;
  }

  public render(props: any): string {
    const elementOfComponentToRendered = createElement(this.Component);
    const propsRendered = {
      children: elementOfComponentToRendered,
      ...props,
    };

    const LayoutElementWithoutCSS: ReactElement = createElement(Layout, propsRendered);
    renderToString(this.sheet.collectStyles(LayoutElementWithoutCSS));

    const LayoutElementWithCSS: ReactElement = createElement(Layout, {
      ...propsRendered,
      styles: this.sheet.getStyleElement(),
    });

    return renderToString(LayoutElementWithCSS);
  }
}

export default SSR;
