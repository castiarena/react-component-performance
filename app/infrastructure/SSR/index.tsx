import React, { ReactElement } from 'react';
import { renderToString } from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';
import { StaticRouterContext } from 'react-router';
import path from 'path';
import { ChunkExtractor } from '@loadable/server';
import { ServerStyleSheet } from 'styled-components';
import Html, { HtmlProps } from '../../components/Html';

const statsFile = path.resolve('./dist/public/loadable-stats.json');

class SSR {
    private sheet: ServerStyleSheet;

    private extractor: ChunkExtractor;

    private readonly Component: React.FC;

    private readonly htmlProps: HtmlProps;

    constructor(Component: React.FC, props: HtmlProps) {
        this.Component = Component;
        this.htmlProps = props;
        this.extractor = new ChunkExtractor({ statsFile, entrypoints: [props.name] });
        this.sheet = new ServerStyleSheet();
    }

    public render(context: StaticRouterContext, location: string = '/'): string {
        const { name, title, lang } = this.htmlProps;
        const { Component } = this;

        const children = this.extractor.collectChunks(this.sheet.collectStyles(
            <StaticRouter location={location} context={context}>
                <Component />
            </StaticRouter>,
        ));

        const styles: ReactElement[] = renderToString(children) && this.sheet.getStyleElement();
        const scripts: ReactElement[] = this.extractor.getScriptElements();
        const links: ReactElement[] = this.extractor.getLinkElements();

        this.sheet.seal();

        return renderToString(
            <Html
                title={title}
                name={name}
                lang={lang}
                styles={styles}
                scripts={scripts}
                links={links}
            >
                {children}
            </Html>,
        );
    }
}

export default SSR;
