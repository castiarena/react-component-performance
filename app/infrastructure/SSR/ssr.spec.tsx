import React from 'react';
import { StaticRouterContext } from 'react-router';
import SSR from '.';
import * as mockStats from './__mocks__/statFile.json';

jest.mock(
    require('path').resolve('./dist/public/loadable-stats.json'),
    () => mockStats,
    { virtual: true },
);

describe('Server side rendering test suite', () => {
    it('should return the html for a component', () => {
        const MockComponent: React.FC = ({ mock }: any) => <div data-mock={mock} />;
        const ssr = new SSR(MockComponent, { title: 'mock title', name: 'mock', lang: 'pt' });
        expect(ssr.render({ mock: true } as StaticRouterContext)).toMatchSnapshot();
    });
});
