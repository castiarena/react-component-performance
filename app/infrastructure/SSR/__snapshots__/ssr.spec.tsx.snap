// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server side rendering test suite should return the html for a component 1`] = `"<html lang=\\"pt\\" data-reactroot=\\"\\"><head><meta charSet=\\"UTF-8\\"/><meta name=\\"viewport\\" content=\\"width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0\\"/><meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\"/><title>mock title</title><link href=\\"https://fonts.googleapis.com/css2?family=Manrope:wght@200;400;800&amp;display=swap\\" rel=\\"stylesheet\\"/><link data-chunk=\\"mock\\" rel=\\"preload\\" as=\\"script\\" href=\\"/public/vendors~mock.bundle.js\\"/><link data-chunk=\\"mock\\" rel=\\"preload\\" as=\\"script\\" href=\\"/public/mock.js\\"/><style data-styled=\\"\\" data-styled-version=\\"5.2.0\\"></style></head><body><main id=\\"mock-wrapper\\"><div></div></main><script id=\\"__LOADABLE_REQUIRED_CHUNKS__\\" type=\\"application/json\\">[]</script><script id=\\"__LOADABLE_REQUIRED_CHUNKS___ext\\" type=\\"application/json\\">{\\"namedChunks\\":[]}</script><script async=\\"\\" data-chunk=\\"mock\\" src=\\"/public/vendors~mock.bundle.js\\"></script><script async=\\"\\" data-chunk=\\"mock\\" src=\\"/public/mock.js\\"></script></body></html>"`;
